### ISSUE TITLE
Add v1alpha2 and Stack CRDs

### ISSUE THREAD

**Issue by kaczyns** [kabanero-io/kabanero-operator#386]
We have received feedback from users that the idea of an Appsody stack and a Kabanero collection being the same thing, is confusing.  A brief overview - a Kabanero collection starts as an Appsody stack, to which we add additional configuration for Kabanero such as pipelines, triggers and image names.  The app developer generally uses Appsody (and an appsody stack) to develop an application, while operations is using Kabanero (and Kabanero collections) to deploy the applications to test and production.

The Kabanero controller should be taught how to read an Appsody stack, with the additional bits and pieces required by Kabanero specified in the Kabanero CR instance, rather than in the Appsody stack.  To this end, we should work towards a v1alpha2 of the Kabanero CRD.  The Collection CRD should be re-named to the Stack CRD to better reflect what it will actually be.

See:
https://github.com/kabanero-io/roadmap/issues/126

**Comment by kaczyns** [comment]
I am proposing the following for the Kabanero v1alpha2 CRD (I'm going to write out a complete CR instance to show all of the fields, I think this is more consumable than listing the schema from the CRD)
```yaml
apiVersion: kabanero.io/v1alpha2
kind: Kabanero
metadata:
  name: kabanero
spec:
  version: "0.6.0"
  stacks: 
    repositories: 
    - name: central
      url: https://github.com/kabanero-io/stacks/releases/download/0.6.0/index.yaml
      skipCertVerification: false  // default is false
    - name: incubator
      url: https://github.com/kabanero-io/stacks/releases/download/0.7.0/index.yaml
      skipCertVerification: false  // default is false
  targetNamespaces:
    - test1
    - deploy1
  github:
    organization: kabanero-io
    apiUrl: http://api.github.com
    teams:
      - admin
      - operations
  pipelines:
    - id: default
      sha256: 0123456789abcdef
      url: https://github.com/kabanero-io/pipelines/releases/download/0.7.0/default.pipeline.tar.gz
  triggers:
    - id: default
      sha256: 0123456789abcdef
      url: https://github.com/kabanero-io/triggers/releases/download/0.7.0/default.triggers.tar.gz
  cliServices:
    version: x.y
    image: kabanero/kabanero-command-line-services:0.6.0
    repository: kabanero
    tag: "0.6.0"
    sessionExpirationSeconds: 1440s
  landing:
    enable: true
    version: x.y
  che:
    enable: false
    cheOperatorInstance:
      cheWorkspaceClusterRole: codewindRole
    kabaneroChe:
      version: x.y
      image: kabanero/kabanero-che:0.6
      repository: kabanero
      tag: "0.6.0"
  events:
    enable: false
    version: x.y
    image: kabanero/kabanero-events:0.1
    repository: kabanero
    tag: "0.1"
  stackController:
    version: x.y
    image: kabanero/kabanero-operator-stack-controller:0.6.0
    repository: kabanero
    tag: "0.6.0"
  admissionControllerWebhook:
    version: x.y
    image: kabanero/kabanero-operator-admission-webhook:0.6.0
    repository: kabanero
    tag: "0.6.0"
```
Note specifically, the added `pipelines` and `triggers` sections.  This CRD does not attempt to give a way to over-ride a pipeline or trigger at the stack version level.  We need to find a way to do this.

**Comment by kaczyns** [comment]
I am proposing the following Stack CRD for v1alpha2.  Same format - this is a fully specified Stack CR instance:
```yaml
apiVersion: kabanero.io/v1alpha2
kind: Stack
metadata:
  name: java-microprofile
spec:
  name: java-microprofile
  versions:
    - repositoryUrl: https://github.com/kabanero-io/stacks/releases/download/0.6.0/index.yaml
      pipelines:
        - id: default
          url: https://github.com/kabanero-io/pipelines/releases/download/0.6.0/default.pipeline.tar.gz
          sha256: 0123456789abcdef
      version: "0.2.15"
      desiredState: ""
      skipCertVerification: false
      images:
        - id: java-microprofile
          image: docker.io/kabanerobeta/java-microprofile:0.2
    - repositoryUrl: https://github.com/kabanero-io/stacks/releases/download/0.7.0/index.yaml
      pipelines:
        - id: default
          url: https://github.com/kabanero-io/pipelines/releases/download/0.7.0/default.pipeline.tar.gz
          sha256: 0123456789abcdef
      version: "0.2.21"
      desiredState: ""
      skipCertVerification: false
      images:
        - id: java-microprofile
          image: docker.io/kabanerobeta/java-microprofile:0.2
```
Note specifically the pipelines and images are now specified in the collection spec.  The Kabanero controller would have read these out of the Kabanero CR instance, and copied them here so the stack controller can process them.

**Comment by groeges** [comment]
@kaczyns Are these CRDs going to be automatically generated or generated and edited by hand?

In the stack CRD where are you intending to get the image information from, ie.
```
      images:
        - id: java-microprofile
          image: docker.io/kabanerobeta/java-microprofile:0.2
```
this is currently added to the index.yaml by the kabanero build scripts (from the collection.index). If we are moving to appsody hubs and using the index.yamls that are generated as part of their builds then this info will not be available?

**Comment by kaczyns** [comment]
@groeges yes I realize that the information will not be available from the appsody hub.... we have to learn how to generate it in the operator and insert it into the CRD, so that the build pipeline can retrieve it and check to see if the stack is active.

**Comment by kaczyns** [comment]
@stephenkinder suggested that the `pipeline` and `triggers` sections live inside the `stacks` section of the Kabanero CRD, since at some point later we may want to introduce pipelines that are not associated with stacks (all of the existing pipelines and triggers are associated with stacks).  I am going to update the first comment to reflect this.

**Comment by groeges** [comment]
@kaczyns We are looking at adding the image element to the generated appsody index.yaml.
It would not be the same format as currently used in the kabanero index yaml but would just be something like:
```
image: docker.io/kabanerobeta/java-microprofile:0.2
```
and will be provided for each stack listed in the index yaml. 
Hopefully this will enable you to provide the same functionality in the operator as you do today.

**Comment by kaczyns** [comment]
Cheers for that Steve.  We appreciate not having to figure out how to build that ourselves.

**Comment by jgawor** [comment]
I'm trying to understand something here. The operator takes in a list of stack repositories and pointers to pipelines, triggers, etc. and creates and maintains the `Stack` resources. Why not drop all of that and just let Champ create and manage the `Stack` resources directly. Each `Stack` object could be specific to a particular version and have Stack specific links. With this, the CLI tools and IDEs could also be extended to "discover" enabled `Stacks` just by using the Kubernetes API. This would eliminate the need for building and maintaining a static index file and make things more dynamic in terms of discovery what `Stacks` are enabled (i.e. per namespace), etc.

**Comment by kaczyns** [comment]
@jgawor Champ can certainly do that if he wants to.  He can leave the repositories section of the Kabanero CR instance empty.  We are going to document this pattern.

The CLI already "discovers" `Collection` instances and will discover `Stack` instances as well (this is how the sync function works).

**Comment by jgawor** [comment]
But why have two different patterns to maintain?

When I look at the proposed `Stack` definition it specifies `repositoryUrl` which points to some `index.yaml` file. I assume that's Appsody's `index.yaml` which could contain multiple stack definitions. And I guess that url was used to load the `Stack` metadata. But if I'm working with a single `Stack`, I might not have an `index.yaml` file and I shouldn't need to create one.

The `kabanero` CLI does but I was thinking about `appsody` CLI.

**Comment by kaczyns** [comment]
You're right - it doesn't make any sense to have the `repositoryUrl` in the `Stack` CR definition.  In the case where the Stack was created by the Kabanero controller, we thought it would be beneficial to know why the Kabanero controller created that stack (what index did it read).  but it probably just makes it more confusing.  @stephenkinder and I just talked about it and we'll take it out.

**Comment by jgawor** [comment]
I would also propose that any references to GitHub be changed to use GitHub API. For example instead of 
```
      pipelines:
        - id: default
          url: https://github.com/kabanero-io/pipelines/releases/download/0.7.0/default.pipeline.tar.gz
          sha256: 0123456789abcdef
```
The `Stack` would support something like:
```
      pipelines:
        - id: default
          github:
             url: https://github.com/kabanero-io/kabanero-pipelines
             revision: 0.5.0 (defaults to master)
```
Or something similar if we want to support using the GitHub release API. The operator would lookup credentials for such endpoints using the same way as [Tekton](https://github.com/tektoncd/pipeline/blob/master/docs/auth.md). This would provide GitHub Enterprise support and potentially add ability to point to the latest release of pipelines `0.2` just like the Stack image does.

This method could also be extended to support various other protocols later on. For example:

```
      pipelines:
        - id: default
          http:
             url: https://github.com/kabanero-io/pipelines/releases/download/0.7.0/default.pipeline.tar.gz
             skipCertVerification: false
             sha256: 0123456789abcdef
```

**Comment by kaczyns** [comment]
@stephenkinder could you comment on Jarek's comment above please, about GHE?  You had asked us in #253 to consider the GHE case holistically across all of the other products (ie appsody).  Does Jarek's proposal fit with what you had in mind?  If we are going to do something like that, then all of the following need to change:
1) stack index yaml in Kabanero CR
2) pipelines tar gz in Stack CR
3) triggers tar gz in Kabanero CR

These changes would also have implications for CLI, landing page, and eventing, since they would need to be able to read the CR and access the files in the same way.

**Comment by stephenkinder** [comment]
@kaczyns @jgawor So I do like the idea of transitioning to _github_ specific syntax suggested by Jarek.   That said, we already have a ton of work to try to get done in 0.6.0.  Do we hate the idea that one could use a URL or GitHub?  In other words, can we add the GitHub syntax as a formal epic and deliver a future milestone?  As Tim says, it is fairly pervasive -- the 0.6.0 cup overfloweth. 

To be clear, the idea is to leave _URL:_ for now and add the GitHub spelling as a separate future change?

**Comment by kaczyns** [comment]
I like that - for now we'd have:
```yaml
spec:
  versions:
    pipelines:
      - id: default
        sha256: abcdef0123456789blahblabhblah
        https:
          url: https://github.com/kabanero-io/pipelines/releases/download/0.7.0/default.pipeline.tar.gz
          skipCertVerification: false
```
and then later we'd add `spec.versions.pipelines.github.blahblah`
Yes?

**Comment by kilnerm** [comment]
@kaczyns, Sorry for weighing in here so late in the day but I think we should consider working this a little differently.

For me the stacks section of the Kabanero CRD should be a pointer to stack index files, preferably individual stack indexes but we can support composite as well (composite may be required for evaluation).

Each of these stack entries should have additional meta-data that can be associated to it, for example desired state. This gives us complete flexibility on a stack by stack basis to control all aspects of its availability within an instance.

If we wanted to have a single entry for multiple versions of a stack we simply move the meta-data into the version specific stack element.

A configuration like this removes any requirement for alternative interfaces for altering the configuration and allows Kabanero to always operate in an autonomous fashion based on it’s CRD.

**Comment by kaczyns** [comment]
Hi @kilnerm - no worries, everyone has an opinion :-)
I'll try to respond to these one by one - I understand where you're coming from, and this is going to continue to evolve, but we need to do something for 0.6.0 and so we might need to put a stake in the ground, and come back to it again later.

The stacks section of the Kabanero CRD can point to individual stack indexes.  We support a list.  If we have a list of 10 indexes, each with 1 stack in them, that's great.  Unless I'm mis-understanding what a stack index actually contains.

We discussed on a call earlier today that we're going to allow the pipelines to be over-ridden for an individual index.  So if someone is maintaining 10 stack indexes, each with 1 stack in them, then they can pick the pipelines that each stack will use by providing the pipeline zip for each index in the Kabanero CR instance.  The desired state becomes irrelevant then - if the user wants it active, they leave the index in the Kabanero CR instance, and if they want it inactive, they delete the index from the Kabanero CR instance.

I'm not sure that I understand the comment about multiple versions of the stack, but given what we've been talking about here, different artifacts for different versions of the same stack can be completely specified in the Kabanero CR instance.

**Comment by kilnerm** [comment]
@kaczyns I thought we covered a lot of this in the call. Like you say we have the ability to revisit this in future releases, and we definitely should, but for now it works.

**Comment by davco01a** [comment]
Did you guys arrive  at where the image info for a given stack will originate from?  CRD or index.yaml....the thread seems to still have some flux around this issue.

A fresh copy of a Kabaner CRD and index.yaml might help

**Comment by stephenkinder** [comment]
@kaczyns I understand your example for image in the Stack CR comes from a direct copy of the data in  the stack hub build/collection hub build logic, but I think when we have multiple versions of the stack, the 2 digit tag is confusing.  When we have `0.2.17` and `0.2.18`, they both have the same image tag which is only 2 digits: `0.2`.  In the container registry, only one of the containers can be tagged `0.2`, it is likely `0.2.18`.   That said, Kabanero needs to be validating containers using the digest field in the registry, and looking up the digest to govern, rather than the tag field.

I think it makes more sense that the image field in the stack, strip any tags off of the image metadata and we should discuss who's the responsibility it is do the registry look ups.  I think it makes sense that the operator add `digest` to the stack versions so that governance code can use digests.

**Comment by kaczyns** [comment]
@stephenkinder This makes sense - I've opened #423 to address stripping the tags.

I agree looking up the digest is the way to go.  I agree that we need to decide who is going to look up the digest.  Also, what to do when the digest changes, since tags are not immutable.  We need to be able to recover from the case where someone changes the tag in the container registry.

**Comment by kilnerm** [comment]
@kaczyns Would it be helpful if the stack build process injected the image digest into the index.yaml?

**Comment by kaczyns** [comment]
@kilnerm Yes I think that would be helpful, then the operator wouldn't need to learn how to talk to docker (or some other registry) to read the digest.  Lets just talk with Steve about it to be sure that this is the right time to get the digest (stack build time).

**Comment by kaczyns** [comment]
The proposed CRDs were merged with #415.  I am going to close this issue.  Please re-open it if anything that we've discussed here (with exception to the image digest) seems missing.
